accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,474 @@
+
+### Byref-like (or `ref struct`) support
+Currently, C# allows one to construct a `TypedReference` to a reference type or value type, but not a byref-like type:
+```cs
+int i = 42;
+TypedReference tr1 = __makeref(i); // okay
+
+Span<int> span = stackalloc int[42];
+TypedReference tr2 = __makeref(span); // Error CS1601: Cannot make reference to variable of type 'Span<int>'
+```
+
+An 8.0 ask from Roslyn is to allow this to compile. See [C# ask for supporting TypedReference + byref-like types](https://github.com/dotnet/roslyn/issues/65255).
+
+Since `TypedReference` internally stores a reference to the **storage location**, and not the actual value or managed-reference-passed-by-value, it effectively supports the `ref\out\in` modifiers. Also, it does this with an implicit `ref` - attempting to use the `ref` keyword is not allowed:
+```cs
+int i = 42;
+TypedReference tr = __makeref(ref i); // error CS1525: Invalid expression term 'ref'
+```
+and not necessary due to `__refvalue`:
+```cs
+int i = 42;
+TypedReference tr = __makeref(i);
+__refvalue(tr, int) = 100;
+Console.WriteLine(i); // 100
+```
+or by using `ref` along with `__refvalue`:
+```cs
+int i = 42;
+TypedReference tr = __makeref(i);
+ChangeIt(tr);
+Console.WriteLine(i); //100
+
+public static void ChangeIt(TypedReference tr)
+{
+    ref int i = ref __refvalue(tr, int);
+    i = 100;
+}
+```
+
+## Byref-like support + simplification constraints
+Below are some simplification constraints that may help with any Roslyn implementation around lifetime rules.
+
+**A `TypedReference` can't reference another `TypedReference`**
+```cs
+TypedReference tr = ...
+TypedReference tr2 = __makeref(tr); //  error CS1601: Cannot make reference to variable of type 'TypedReference'
+```
+
+This is similar to trying to make a `TypedReference` to a `Span<int>` above -- both are byref-like types.
+
+However, this limitation for `TypedReference-to-TypedReference` does **not** need to be removed for the goals in this design if it helps with simplifying lifetime rules.
+
+**Just support what is allowed today.** No new capabilities are expected; reflection invokes existing methods which must have already been compiled according to existing rules:
+```cs
+internal class Program
+{
+    static void Main()
+    {
+        // Case 1
+        MyRefStruct rs = default;
+        TypedReference tr = __makeref(rs._span); // Assume we allow instead of CS1601
+        ChangeIt(tr);
+
+        // Case 2
+        Span<int> heap = new int[42];
+        Span<int> stack = stackalloc int[42];
+        CallMe(ref stack, heap); // okay
+        CallMe(ref heap, stack); // CS8350 as expected
+    }
+
+    public static void ChangeIt(TypedReference tr)
+    {
+        // This compiles today:
+        ref Span<int> s = ref __refvalue(tr, Span<int>);
```

Comment:
I think the easiest path forward for `TypedReference` safely holding a `Span<int>` is for this to turn into an `unsafe` operation. This is where the danger really sets in, unless we significantly complicate the existing lifetime rules, hence this is where `unsafe` should go. Essentially this is when the rules break down if we allow capturing `Span<int>` safely. 

Note: specifically I mean when `ref __refvalue(expr, T)` and `T` is a `ref struct`.